### ISSUE TITLE
Use recyclable & paginated buffers for search hits

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#99590]
Today `SearchHit#source` is a plain `BytesReference` which is probably inappropriate since hits are occasionally large and often quite short-lived. For instance, if fetching hits from a remote node, the coordinator will read this field into a single `new byte[]` allocated within here:

https://github.com/elastic/elasticsearch/blob/19257125b19ff05a6b014f37693f9417ca579626/server/src/main/java/org/elasticsearch/search/SearchHit.java#L152

Those allocations can be humongous and will impose quite some load on the GC. We should instead consider move to using recyclable and paginated buffers for search hits. In terms of network reads that means using `StreamInput#readReleasableBytesReference()` and anywhere we're allocating it locally we should use the node's `BigArrays`. Unfortunately that has quite wide-reaching consequences, since it would mean that `SearchHit` must now become `Releasable`, as must all its owning classes, so that the buffers can be recycled once they're no longer in use.

This has substantial overlap with https://github.com/elastic/elasticsearch/issues/89656 since a large part of the work involved in tracking this memory usage for circuit-breaking purposes would also require us to add such a lifecycle to `SearchHit` and its owners.

**Comment by iverase** [comment]
Those source bytes are normally compressed, so the situation becomes more difficult when we decompress it, as it generates yet another bigger plain BytesReference:

https://github.com/elastic/elasticsearch/blob/795e07c021fea92b566ad7be44268e98cbaeb1c7/server/src/main/java/org/elasticsearch/search/SearchHit.java#L379

**Comment by DaveCTurner** [comment]
That's true, but it isn't hard to decompress a recyclable `BytesReference` into another recyclable `BytesReference` too as long as we can add the plumbing to make the recycler available where it's needed.

**Comment by iverase** [comment]
What will be harder is that those plain bytes come all the way from lucene: 

https://github.com/apache/lucene/blob/008a0d420635239570e09e0862a357f512781c60/lucene/core/src/java/org/apache/lucene/codecs/lucene90/compressing/Lucene90CompressingStoredFieldsReader.java#L243

I hope we can tell lucene to give as a reference on the shape of a DataInput or similar (like in this proposal https://github.com/apache/lucene/pull/12460)?

**Comment by DaveCTurner** [comment]
> I hope we can tell lucene to give as a reference on the shape of a DataInput or similar

Yeah that's certainly another place worth fixing. But the problems in this area are particularly noticeable on coordinating nodes today, so fixing this in the transport layer will already help. And the work to give `SearchHit` a proper lifecycle will pay dividends in various other ways too.

**Comment by benwtrent** [comment]
@original-brownbear can we close this?

**Comment by javanna** [comment]
What's left to do (credits to @original-brownbear ): remove all remaining uses of SearchHit#asUnpooled  if we wanted to be thorough + make Lucene read sources to pooled buffers as well

**Comment by mayya-sharipova** [comment]
We had another use case of OOM happened.
Huge search responses with 1000 hits, each hit has a source of 198Kb.  
But OOM happened when ES tried to serialize search hits:

<img width="1070" alt="Image" src="https://github.com/user-attachments/assets/e8f937c1-6190-433d-a20f-0642b743891e" />
